### ISSUE TITLE
Writing a cupy array caster for pybind11 bindings

### ISSUE THREAD

**Issue by SimeonEhrig** [ComputationalRadiationPhysics/student_project_python_bindings#17]
During the development of PR #15 we recognized, that we cannot use the numpy array support of pybind11 for cupy arrays. The reason is, that both arrays uses different apis.

* **cupy** : https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html#cuda-array-interface
* **numpy**: https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.interface.html#__array_interface__

Therefore we need to implement a [custom pybind11 caster](https://pybind11.readthedocs.io/en/stable/advanced/cast/custom.html) to support a nice interface for the C++ and Python side.

Please implement the custom caster. The C++ interfaces should look like this:

```C++
void compute(mycupy::arrray_t<float> x, mycupy::arrray_t<float> y){
  kernel<<<1,1>>>(x.ptr, y.ptr);
}
```

This documentation contains useful information about, how a Python class is organized internal: https://docs.python.org/3/reference/datamodel.html

**Comment by SimeonEhrig** [comment]
TLDR; the bindings should return the `cupy array` directly, because the return value is a copy each time.

**Comment by SimeonEhrig** [comment]
**Update:** During the development of the cupy caster, we got new experience using pybind11, which changes some of the design decisions.
- **Python to C++:** By default, pybind11 passes arguments by copy from the Python to the C++ side. With a custom caster, it is also possible to pass a reference via pointer. But the breaks the expected behavior of the pybind11 api. Therefor we introduce the `Cupy_Ref`class, which is a reference to a cupy array and pass this wrapper class by reference to the C++ side.
- `C++ to Python:` For the other way, we can use directly a `cupy array`, because the memory needs to be allocate each time in the C++ function and the ownership will be moved to the Python side after leaving the C++ function.
